lib/cog/pipeline.ex

```diff
@@ -0,0 +1,337 @@
+  alias Cog.Messages.ProviderRequest
+  alias Cog.Models.EctoJson
+  alias Cog.Pipeline.PrepareError
+  alias Cog.Pipeline.{CommandResolver, DataSignal, DoneSignal, ExecutionStageSup, ErrorSinkSup,
+                      InitialContext, InitialContextSup, OutputSinkSup, PermissionsCache}
+  alias Cog.Queries.User, as: UserQueries
+  alias Cog.Repo
+  alias Piper.Command.Ast
+  alias Piper.Command.Parser
+  alias Piper.Command.ParserOptions
+
+  use GenServer
+
+  require Logger
+
+  @type output_policy :: :adapter | :owner | :adapter_owner
+
+  defstruct [:policy, :owner, :request, :stages, :token, :conn, :started, :status]
+
+  @doc """
+  Starts a new pipeline and prepares it to begin
+  processing
+
+  ## Options
+  * `:request` - Originating `Cog.Messages.ProviderRequest`. Required.
+  * `:owner` - Pid of pipeline owner. Optional.
+  * `:output_policy` - Controls where to send output. Valid values for this option are:
+    * `:adapter` - Pipeline will send output via `Cog.Chat.Adapter` (default)
+    * `:owner` - Pipeline will forward output to the owning process. Requires `:owner` option.
+    * `:adapter_owner` - Pipeline sends output via `Cog.Chat.Adapter` _and_ forwards
+       a copy to the owning process. Requires `:owner` option.
+  """
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, [opts])
+  end
+
+  @doc """
+  Tells a prepared pipeline to begin executing
+  """
+  def run(pipeline) do
+    GenServer.call(pipeline, :run, :infinity)
+  end
+
+  @doc """
+  Notifies pipeline when processing is finished
+  """
+  def notify(pipeline) do
+    GenServer.cast(pipeline, :teardown)
+  end
+
+  def init([opts]) do
+    try do
+      policy = Keyword.get(opts, :output_policy, @default_output_policy)
+      request = Keyword.fetch!(opts, :request) |> sanitize_request
+      owner = if policy in [:owner, :adapter_owner] do
+        Keyword.fetch!(opts, :owner)
+      else
+        nil
+      end
+      case ConnectionSup.connect() do
+        {:ok, conn} ->
+          {:ok, %__MODULE__{policy: policy, owner: owner,
+                            request: request, conn: conn,
+                            token: Tokens.new(), status: :running}}
+        error ->
+          Logger.error("Failed to connect pipeline #{request.id} to message bus: #{inspect error}")
+          {:stop, error}
+      end
+    rescue
+      e in KeyError ->
+        {:stop, {:missing_option, e.key}}
+    end
+  end
+
+  def handle_call(:run, _from, state) do
+    state = %{state | started: DateTime.utc_now()}
+    case fetch_user(state) do
+      {:ok, user} ->
+        with {:ok, parsed, destinations} <- parse(user, state),
+             {:ok, perms} <- PermissionsCache.fetch(user),
+               user_json <- EctoJson.render(user) do
+          start_pipeline(parsed, destinations, user_json, perms, state)
+        else
+          {:error, {:parse_error, message}} -> start_error_pipeline({:error, :parse_error, message}, user, state)
```

Comment:
Can we put these clauses on a separate line from the conditions for readability?